include/robot_interfaces/robot_logger.hpp

```diff
@@ -0,0 +1,331 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2018-2019, Max Planck Gesellschaft
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <fstream>
+#include <iostream>
+
+#include <chrono>
+
+#include <Eigen/Eigen>
+
+#include <mpi_cpp_tools/basic_tools.hpp>
+#include <mpi_cpp_tools/dynamical_systems.hpp>
+#include <mpi_cpp_tools/math.hpp>
+#include <mpi_cpp_tools/infix_iterator.h>
+
+#include <real_time_tools/process_manager.hpp>
+#include <real_time_tools/thread.hpp>
+#include <real_time_tools/threadsafe/threadsafe_timeseries.hpp>
+#include <real_time_tools/timer.hpp>
+
+#include <robot_interfaces/loggable.hpp>
+
+namespace robot_interfaces
+{
+  /**
+   * @brief To log data from *any* robot (real, simulated, fake).
+   *
+   * The logger logs the timestamp, the time index, and the values of every
+   * observation, action, and status variable. Further loggable data structures
+   * can be added to the logger, which *must* derive from class Loggable.
+   *
+   * @tparam Action
+   * @tparam Observation
+   * @tparam Status
+   */
+template <typename Action, typename Observation, typename Status>
+class RobotLogger
+{
+public:
+
+    static_assert(std::is_base_of<Loggable, Action>::value,
+                  "Action must derive from Loggable");
+    static_assert(std::is_base_of<Loggable, Observation>::value,
+                  "Observation must derive from Loggable");
+    static_assert(std::is_base_of<Loggable, Status>::value,
+                  "Status must derive from Loggable");
+
+    /*
+    * Currently, the level of generalisability of the logger is that we know
+    * the following four structures exist for *any* robot whose data is to be
+    * logged.
+    */
+
+    Action applied_action;
+    Action desired_action;
+    Observation observation;
+    Status status;
+
+    std::shared_ptr<robot_interfaces::RobotData<Action, Observation, Status>>
+        logger_data_;
+
+    std::vector<std::string> header_;
+
+    int block_size_;
+    long int index_;
+
+    bool stop_was_called_;
+
+    std::ofstream output_file_;
```

Comment:
As the file is always closed at the end of each function where it is opened, I think it would be better to define this variable locally in each function instead of having it as a class member.

Reply:
I will look into this later.